experimental/ir/ir_value.go

```diff
@@ -0,0 +1,479 @@
+//
+// They are all together here so it's easier to remember to delete them at the
+// same time, instead of putting //nolint on each function.
+//
+// TODO: Delete this.
+var _ = []any{
+	newScalar[int32], appendScalar[int32],
+	newMessage, appendMessage,
+	MessageValue.insert,
+}
+
+// Value is an evaluated expression, corresponding to an option in a Protobuf
+// file.
+type Value struct {
+	withContext
+	raw *rawValue
+}
+
+// rawValue is a [rawValueBits] with field information attached to it.
+type rawValue struct {
+	// The expression that this value was evaluated from.
+	//
+	// If this is an entry in a [MessageValue], this will be an ast.FieldExpr.
+	expr ast.ExprAny
+
+	// The AST node for the path of the option (compact or otherwise) that
+	// specifies this value. This is intended for diagnostics.
+	//
+	// For example, the node
+	//
+	//  option a.b.c = 9;
+	//
+	// results in a field a: {b: {c: 9}}, which is four rawFieldValues deep.
+	// Each of these will have the same optionPath, for a.b.c.
+	optionPath ast.Path
+
+	// The field that this value sets. This is where type information comes
+	// from.
+	//
+	// NOTE: we steal the high bit of the pointer to indicate whether or not
+	// bits refers to a slice. If the pointer part is negative, bits is a
+	// repeated field with multiple elements.
+	field ref[rawField]
+	bits  rawValueBits
+}
+
+// rawValueBits is used to represent the actual value for all types, according to
+// the following encoding:
+//  1. All numeric types, including bool and enums. This holds the bits.
+//  2. String and bytes. This holds an intern.ID.
+//  3. Messages. This holds an arena.Pointer[rawMessage].
+//  4. Repeated fields with two or more entries. This holds an
+//     arena.Pointer[[]rawValueBits], where each value is interpreted as a
+//     non-array with this value's type.
+//     This exploits the fact that arrays cannot contain other arrays.
+//     Note that within the IR, map fields do not exist, and are represented as
+//     the repeated message fields that they will ultimately become.
+type rawValueBits uint64
+
+// AST returns the expression that evaluated to this value.
+func (v Value) AST() ast.ExprAny {
+	if v.IsZero() {
+		return ast.ExprAny{}
+	}
+
+	if field := v.raw.expr.AsField(); field.IsZero() {
+		return field.Value()
+	}
+
+	return v.raw.expr
+}
+
+// OptionPath returns the AST node for the path of the option (compact or not)
+// that caused this value to be set.
+func (v Value) OptionPath() ast.Path {
+	if v.IsZero() {
+		return ast.Path{}
+	}
+	return v.raw.optionPath
+}
+
+// FieldAST returns the field expression that sets this value in a message
+// expression, if it is such a value.
+func (v Value) FieldAST() ast.ExprField {
+	if v.IsZero() {
+		return ast.ExprField{}
+	}
+
+	return v.raw.expr.AsField()
+}
+
+// Type is the element type of this value. For arrays, this is the type of the
+// element; Protobuf does not have a concept of an "array type".
```

Comment:
How does a caller know if it's an array type? While the underlying storage could always use a negative value for the field (even if the actual literal in the AST has exactly one value), we seem to lose that knowledge in this type's API: there is no accessor to ask if it's an array and just counting the number of `v.Elements()` doesn't work since a singular value and an array of one value would both reutrn the same thing.